src/main/java/com/google/sps/DataServlet.java

```diff
@@ -212,33 +271,36 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         && (diff == null || diff.getMutationList().size() == 0)) {
       response.setHeader(
           "serverMessage",
-          "The searched node does not exist in this graph, so nothing is shown. However, it is"
-              + " mutated at some other step. Please click next or previous to navigate to a graph"
-              + " where this node exists.");
+          "The searched node/token does not exist in this graph, so nothing is shown. However, it"
+              + " is mutated at some other step. Please click next or previous to navigate to a"
+              + " graph where this node exists.");
     }
     // The searched node exists but is not mutated in the current graph
     if (truncatedGraph.nodes().size() != 0
-        && !(mutationNumber == -1 && nodeNameParam.length() == 0)
-        && filteredMutationIndices.indexOf(mutationNumber) == -1) {
+        && !(mutationNumber == -1 && nodeNames.size() == 0 && tokenParam.length() == 0)
+        && filteredMutationIndices.indexOf(mutationNumber) == -1
+        && (diff == null || diff.getMutationList().size() == 0)) {
       response.setHeader(
           "serverMessage",
-          "The searched node exists in this graph! However, it is not mutated in this graph."
-              + " Please click next or previous if you wish to see where it was mutated!");
+          "The searched node/token exists in this graph. However, it is not mutated in this"
+              + " graph. Please click next or previous if you wish to see where it was"
+              + " mutated!");
     }
+
     // There is a diff between the previously-displayed graph and the current graph but
     // no mutations in it only mutate on-screen nodes
     if (filteredMutationIndices.indexOf(mutationNumber) != -1
-        && diff != null
-        && diff.getMutationList().size() == 0) {
+        && filteredDiff != null
+        && filteredDiff.getMutationList().size() == 0) {
       response.setHeader(
           "serverMessage",
           "The desired set of nodes is mutated in this graph but your other parameters (for eg."
               + " radius), limit the display of the mutations. Please try increasing your radius"
               + " to view the mutation.");
     }
-
     graphJson =
-        Utility.graphToJson(truncatedGraph, filteredMutationIndices, diff, mutList.size(), queried);
+        Utility.graphToJson(
+            truncatedGraph, filteredMutationIndices, filteredDiff, mutList.size(), queried);
```

Comment:
Should we return queried or queriedNext?

Reply:
I think it is queriedNext, right?